src/ImageSharp.Web/DependencyInjection/ImageSharpBuilderExtensions.cs

```diff
@@ -96,6 +96,33 @@ public static IImageSharpBuilder SetCache(this IImageSharpBuilder builder, Func<
             return builder;
         }
 
+        /// <summary>
+        /// Sets the given <see cref="ICacheKey"/> adding it to the service collection.
+        /// </summary>
+        /// <typeparam name="T">The type of class implementing <see cref="ICacheKey"/> to add.</typeparam>
+        /// <param name="builder">The core builder.</param>
+        /// <returns>The <see cref="IImageSharpBuilder"/>.</returns>
+        public static IImageSharpBuilder SetCacheKey<T>(this IImageSharpBuilder builder)
+            where T : class, ICacheKey
+        {
+            var descriptor = new ServiceDescriptor(typeof(ICacheKey), typeof(T), ServiceLifetime.Singleton);
```

Comment:
Looks like the default implementations of `ICacheKey` don't have any state. In that case it's better to use `Transient` since it eliminates a lookup in the DI entries. It will just emit an efficient `new`.

Reply:
That's interesting. I always thought that would be the opposite. That the lookup would be as expensive for both and that since it was stateless we can use the singleton and avoid the per request allocation.

Reply:
I'll ask a few gods to confirm.

Reply:
It depends on how efficient allocating a new instance of T each time is vs retrieving the same instance.

When you're calling `sp.GetService(typeof(ICacheKey))`, the provider will have to search [a ConcurrentDictionary](https://github.com/dotnet/runtime/blob/479ee518f32cfaf53285b915265ccead9b6b13c9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L18) for an `ICacheKey` service "call site" regardless. This happens for any service type no matter the lifetime when resolved directly with `GetService`.

A transient lifetime means there's not an additional lookup into [another ConcurrentDictionary](https://github.com/dotnet/runtime/blob/479ee518f32cfaf53285b915265ccead9b6b13c9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs) to try to get an already-resolved service. Singleton services are simply resolved from a special "root scope", so they're not all that different from scoped services as far as this lookup goes.

If you have a service with a constructor like `ServiceA(ICacheKey)` and you call `sp.GetService(typeof(ServiceA))`, then the service provider will lookup `ServiceA` from the call site dictionary, but it will not need to lookup the `ICacheKey` service from the same dictionary. `ServiceA`'s dependencies are already embedded in the call site.


If the `ICacheKey` service is transient, `ServiceA` will be constructed by just newing up the ICacheKey implementation type: `new ServiceA(new T())`. If the `ICacheKey` service is singleton or scoped, it will be more like: `new ServiceA(
ResolvedServices.GetOrAdd(_ => new T())`.

The reality is a little more complicated because even with ConcurrentDictionary, there needs to be a lock to ensure the same service doesn't get double initialized withing the same scope. There's also some additional logic to track disposables. You can see the real implementation [here](https://github.com/dotnet/runtime/blob/479ee518f32cfaf53285b915265ccead9b6b13c9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs#L295-L322). You'll notice that the `else` branch of the `if (callSite.Cache.Location == CallSiteResultCacheLocation.Scope)` is a lot simpler.

Reply:
I've simply used the same lifetime as in all the other extension methods 😇

Reply:
Thanks for all the information @halter73 I think in our case any performance difference would be virtually unmeasurable. We'll just stick with the same lifetime for convenience.